File: Java/opentraceability/src/main/java/opentraceability/utility/UOM.java
package opentraceability.utility;

import opentraceability.OTLogger;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

public class UOM {
    Object _locker = new Object();
    public String Name = "";
    public String Abbreviation = "";
    public String UnitDimension = "";
    public String SubGroup = "";
    public String UNCode = "";
    public double A = 0.0;
    public double B = 0.0;
    public double C = 0.0;
    public double D = 0.0;
    public double Offset = 0.0;

    public UOM() {}

    public UOM(UOM uom) {
        this.Abbreviation = uom.Abbreviation;
        this.Name = uom.Name;
        this.UnitDimension = uom.UnitDimension;
        this.UNCode = uom.UNCode;
        this.SubGroup = uom.SubGroup;
        this.Offset = uom.Offset;
        this.A = uom.A;
        this.B = uom.B;
        this.C = uom.C;
        this.D = uom.D;
    }

    public UOM(JSONObject juom) throws Exception {
        this.A = 0.0;
        this.B = 1.0;
        this.C = 1.0;
        this.D = 0.0;

        this.Name = juom.has("name") ? juom.getString("name") : null;
        this.UNCode = juom.has("UNCode") ? juom.getString("UNCode") : null;
        this.Abbreviation = juom.has("symbol") ? juom.getString("symbol") : null;
        this.UnitDimension = juom.has("type") ? juom.getString("type") : null;

        Offset = juom.has("offset") ? juom.getDouble("offset") : 0.0;

        Object multiplier = juom.has("multiplier") ? juom.get("multiplier") : null;
        if (multiplier instanceof String) {
            String multiplierString = (String)multiplier;
            int numerator = Integer.parseInt(multiplierString.split("/")[0]);
            int denominator = Integer.parseInt(multiplierString.split("/")[1]);
            B = numerator;
            C = denominator;
        } else {
            B = Double.parseDouble(multiplier.toString());
        }
    }

    public static boolean isNullOrEmpty(UOM uom) {
        return uom == null || uom.Abbreviation.isEmpty();
    }

    private static List<UOM> uomList = null;
    private static final ReentrantLock uomListLock = new ReentrantLock();

    public void CopyFrom(UOM uom) {
        this.Abbreviation = uom.Abbreviation;
        this.Name = uom.Name;
        this.UnitDimension = uom.UnitDimension;
        this.UNCode = uom.UNCode;
        this.SubGroup = uom.SubGroup;
        this.Offset = uom.Offset;
        this.A = uom.A;
        this.B = uom.B;
        this.C = uom.C;
        this.D = uom.D;
    }

    public boolean isBase() {
        return A == 0.0 && B == 1.0 && C == 1.0 && D == 0.0;
    }

    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof UOM)) {
            return false;
        }
        UOM otherUOM = (UOM) other;
        return UNCode.equals(otherUOM.UNCode);
    }

    public int hashCode() {
        return UNCode.hashCode();
    }

    public String key() {
        return Abbreviation;
    }

    public double convert(double value, UOM to) {
        double valueBase = toBase(value);
        double valueNew = to.fromBase(valueBase);
        return valueNew;
    }

    public double toBase(double value) {
        double baseValue = ((A + B * value) / (C + D * value)) - Offset;
        return baseValue;
    }

    public double fromBase(double baseValue) {
        double value = ((A - C * baseValue) / (D * baseValue - B)) + Offset;
        return value;
    }

    public String toString() {
        return Name + " [" + Abbreviation + "]";
    }
}

File: Java/OpenTraceability/src/main/opentraceability/utility/uom.kt
package utility

import com.intellij.json.psi.JsonObject
import OTLogger


import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

 class UOM {
     val Name: String = ""
     val Abbreviation: String = ""
     val UnitDimension: String = ""
     val SubGroup: String = ""
     val UNCode: String = ""
     val A: Double= 0.0
     val B: Double= 0.0
     val C: Double= 0.0
     val D: Double = 0.0
     val Offset: Double = 0.0

    companion object {
        private val uomListLock = ReentrantLock()
        private var uomList: List<UOM>? = null

        fun lookUpFromUNCode(unCode: String): UOM {
            return uomListLock.withLock {
                val uom = getUOMList().find { it.UNCode == unCode }
                return uom ?: UOM()
            }
        }

        fun isNullOrEmpty(uom: UOM?): Boolean {
            return uom == null || uom.Abbreviation.isNullOrEmpty()
        }

        fun parseFromName(name: String): UOM {
            try {
                if (name.isNullOrEmpty()) {
                    throw IllegalArgumentException("name")
                }

                val uom = getUOMList().find { it.Name == name }
                    ?: getUOMList().find { it.UNCode.equals(name, ignoreCase = true) }
                    ?: throw Exception("Failed to parse UOM")

                return UOM(
                    uom.Name,
                    uom.Abbreviation,
                    uom.UnitDimension,
                    uom.SubGroup,
                    uom.UNCode,
                    uom.A,
                    uom.B,
                    uom.C,
                    uom.D,
                    uom.Offset
                )
            } catch (ex: Exception) {
                OTLogger.error(ex)
                throw ex
            }
        }

        fun getUOMList(): List<UOM> {
            if (uomList == null) {
                // Initialize the list if it's null
                uomList = loadUOMList()
            }
            return uomList!!
        }

        fun loadUOMList(): List<UOM> {
            // Perform the loading of UOMs from the appropriate source
            // and return the list of UOM objects
            // ...
            return emptyList()
        }
    }

    fun isBase(): Boolean {
        return A == 0.0 && B == 1.0 && C == 1.0 && D == 0.0
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is UOM) {
            return false
        }
        return UNCode == other.UNCode
    }

    override fun hashCode(): Int {
        return UNCode.hashCode()
    }

    val key: String
        get() = Abbreviation

    fun convert(value: Double, to: UOM): Double {
        val valueBase = this.toBase(value)
        val valueNew = to.fromBase(valueBase)
        return valueNew
    }

    fun toBase(value: Double): Double {
        val baseValue = ((A + B * value) / (C + D * value)) - Offset
        return baseValue
    }

    fun fromBase(baseValue: Double): Double {
        val value = ((A - C * baseValue) / (D * baseValue - B)) + Offset
        return value
    }

    override fun toString(): String {
        return "$Name [$Abbreviation]"
    }
}
